lib/Importer/Caffe2ModelLoader.cpp

```diff
@@ -524,6 +525,65 @@ llvm::Error Caffe2ModelLoader::loadConvQuantized(const caffe2::OperatorDef &op,
   return llvm::Error::success();
 }
 
+llvm::Expected<bool>
+Caffe2ModelLoader::foldOperator(const caffe2::OperatorDef &op) {
+  const unsigned numInputs = op.input_size();
+  const std::string &typeName = op.type();
+  llvm::SmallVector<NodeValue, 4> inputs;
+  inputs.reserve(numInputs);
+  for (unsigned i = 0; i < numInputs; i++) {
+    NodeValue in;
+    ASSIGN_VALUE_OR_RETURN_ERR(in, getNodeValueByName(op.input(i)));
+    inputs.push_back(in);
+  }
+
+  if (!isConstantFoldable(inputs, typeName)) {
+    return false;
+  }
+
+  // Create a temporary lightweight loader and execution engine.
+  PlaceholderBindings binding_temp;
+  ExecutionEngine ee_temp;
+  Module &mod_temp = ee_temp.getModule();
+  Function *pfunc_temp = mod_temp.createFunction("temp");
+  Caffe2ModelLoader loader_temp(*pfunc_temp, nullptr);
+
+  // Create a temporary graph with constant inputs, current operator and add
+  // save nodes. Recheck if even save nodes are necessary.
+  for (unsigned i = 0; i < numInputs; i++) {
+    Node *pn = inputs[i].getNode();
+    const Tensor &data = (llvm::dyn_cast<Constant>(pn))->getPayload();
+    Constant *pConst = mod_temp.createConstant(op.input(i), data);
```

Comment:
This copy of constant payload is temporary and should get freed upon exiting const fold. I chose to not move the original payload with an intention of not disturbing the original graph. Move would have required me to move back the constant to the original graph, and with RETURN_IF_ERR in the code, I didn't want to deal with that at that time. 

I toyed with the idea of writing a compile_light with optimizations turned off but apparently some operator depended on the optimization so I reverted back to the original compile. 

I am not sure why it is a problem if the constant is quantized, which anyway would have quantized in the later pass. I think during the later later pass it probably won't touch the already quantized constant. Pardon me, I haven't tried quantized networks hence my limited understanding on this example. Still I think since this is an optional feature, a user can always choose to not use it if some side-effects are seen. Also if there is a case encountered we can update the constantFoldable logic to indicate something as non-foldable.

Reply:
@jfix71  I actually tried to reuse the parent module by create a new temporary function let's call it "fold". My expectation was that this way I don't have to copy the constants and it could directly have access to the constants from the parent module. I view constants as a shared resource between functions created out of the same module. However, when the temporary function "fold" was optimized as part of compile it was started deleting the constants of the modules even though they were not part of the "fold" function or graph. Shouldn't it delete only the dead constants which are part of the current function, instead of deleting the modules of the parent module? Either this is a bug or my understanding is wrong.

Reply:
>However, when the temporary function "fold" was optimized as part of compile it started deleting the constants of the parent module even though they were not part of the "fold" function or graph. 

Your understanding is correct (or at least matches mine 🙂) -- this sounds wrong/broken. This is similar to what #3161 does -- create a new Function in the Module which references the same Constants and run it. How can you tell it's deleting the Constants, what kind of error is it throwing?

Also, I think #3161 is a more generalize implementation than this PR. I believe if we're able to get that one to work we can call into it from the Loaders like this one does. We would still need to add calls for it of course.

Reply:
@jfix71 From constant folding at loader perspective I find this implementation to be much simpler and general enough.:-) I am hoping if loader can take care of this upfront, a lot less will percolate to the node level. I am not sure if it can be compared with #3161 as that one does not use the loader and probably will be called once the loading is finished. 

I am speculating that the deletion happens because I call the folding function while I am still loading the graph. All the uses of constants are still not currently marked in the original graph as it's only partially loaded. So it's possible when DCE is called for the temporary graph from the same module, it starts deleting the constants. I will be writing a compile_light, which I guess will solve the problem, unless somebody is already doing this.

Reply:
@jfix71 Have added a FunctionState to the objects of type Function to indicate if the Function has been created/loaded/compiled/etc. In DCE introduced a check for deleteConstant to ensure that constants are deleted only if all the Functions in the parent modules  are already loaded. Now I don't have to copy the constants. I think in general it will be good to have state of Functions set to something meaningful.